### DISCUSSION TITLE
chapter_computational_complexity/performance_evaluation/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#11] [Announcements]
# chapter_computational_complexity/performance_evaluation/

Your first book to learn Data Structure And Algorithm.

https://www.hello-algo.com/chapter_computational_complexity/performance_evaluation/

**Comment by hsamgle** [comment]
> 数据结构与算法追求 “运行地快、内存占用少” ，而如何去评价算法效率则是非常重要的问题

运行"地"快，是不是有错别字？

**Reply by krahets** [reply]
感谢指正！（可以点击页面右上角的“铅笔”图标，直接修改原文内容，谢谢！）

**Reply by nobt854** [reply]
感觉“运行速度快、占用内存少”更优

**Reply by 1931668642** [reply]
二者不可兼得

**Comment by sakkyakei** [comment]
有个小建议：
因为作者希望初学者初步了解一下计算复杂度，这里是不是可以放一些您觉得比较适合小白的科普文呢？

**Reply by krahets** [reply]
哈喽，下一节“时间复杂度”和“空间复杂度”即计算复杂度，有做详细介绍～

**Reply by Wenwen555** [reply]
个人认为作者已经深入浅出的讲解了复杂度的概念啦，尽管没有一个很具体的例子～

**Comment by husenlinwei** [comment]
这一句“复杂度分析评估的是算法运行效率随着输入数据量增多时的增长趋势。” 貌似可以润色成 “复杂度分析旨在评估随着输入数据量增加，算法运行效率所呈现的增长趋势。” ？

**Reply by krahets** [reply]
已润色该定义～

**Reply by ljl312869** [reply]
应该去掉增长 有些算法运行效率和数据量是无关的 是一个常量

**Comment by yicijohncy** [comment]
“它描述了随着输入数据大小的增加，算法执行所需时间和空间的增长趋势”这个定义虽然抽象，但也挺好理解。

前面说过“随着输入数据量的变化，算法会表现出不同的效率”。

也就是说，**数据量**是「算法执行效率」的重要影响因素。

而**时间**和**空间**又是我们重要的消费资源。

把**数据量**看做 x 值，算法计算 x 值所需**时间**或**空间**当做 y 值。

计算出来的数据，就是一个横坐标的点（x，y）。

把不同的点，连接起来，就会是一条曲线。

不同的算法，就会有属于自己的不同“曲线”。

通过观察这条曲线的“陡峭”程度，也就是时间或空间增长的“快慢”，从而评价这个算法的性能。

**Reply by krahets** [reply]
栩栩如生！

**Reply by 7CkShy** [reply]
那已知一系列的x，怎么得到对应的y值呢？通过实际计算机运行统计出来？

**Comment by xx10n31y** [comment]
> 如果你仍对复杂度的概念感到困惑，无须担心，我们会在后续章节中详细介绍。

**哈哈, 确实感到些许困惑, 可能这就是理论知识让人略感抽象**

**Comment by liangguitao** [comment]
Day 1 : LGT  总结 ：理论估算不同的算法  最后我们都可以用（x,y）抛物线的形式来表示，不同的算法，随着数据量的增加（x改变），那么时间或者空间也可能发生改变（y改变）

**Comment by zerlindali** [comment]
我连复杂度都看不懂，咋搞

**Reply by krahets** [reply]
可以理解个大概，先看后面几章，回头再看这里

**Comment by mix8889** [comment]
嗯，老板
真的写的很好很赞！

**Comment by yalishiduozhang** [comment]
太强啦！

**Comment by UNUKyfmt** [comment]
太赞了

**Comment by Rinato-C417** [comment]
Completed

**Reply by krahets** [reply]
Hi，建议只在每章的首节打卡，
欢迎在内容章节多分享见解和提问噢～

**Comment by yujingzhou** [comment]
‧ 时间效率:算法运行速度的快慢。
这个定义不太准确。更准确的定义应该是

‧ 时间效率:算法运行的计算次数。

**Reply by krahets** [reply]
“时间效率”整体上指的就是算法在“时间”上的快慢，而非运行计算次数。

即使算法 A 比算法 B 的计算次数更多，但算法 A 仍然可能比算法 B 更快（算法 B 的单元操作的平均计算时间长）。

**Comment by gofunbox** [comment]
我的智商都看得懂

**Comment by NgxuAnGMH** [comment]
抛去拗口的概念，采取两种方式实现同一个功能，但是不知道它们的优劣，直接为了这么小的功能进行测试（甚至是整个系统的测试），一是不切实际（资源有限），二是未必能屏蔽所有干扰（环境完全一致）。也因此考虑用复杂度来衡量，估算“时间”和“空间”各自的开销，是一种完全从代码层面（嵌套循环/所用内存多少）就可以估算的指标。

**Comment by DayDayYES** [comment]
下面有一个关于时间效率的讨论，作者认为：“时间效率”整体上指的就是算法在“时间”上的快慢，而非运行计算次数。还有人则说： 时间效率指算法运行的计算次数。

**Reply by krahets** [reply]
是的，原因很直接：我们提出时间效率这个概念，最终想要关心的是“时间”上的快慢，而不是“执行次数”。

执行次数更多和时间复杂度关联，它是反映时间效率的有效指标，但很有可能给出错误的结论。我们有时会遇到一种情况：算法 A 比算法 B 的时间复杂度更高（更差），但反而在给定数据下运行地更快（时间更短、效率更高）。一个典型的例子是插入排序 vs. 归并排序在数据量较小时的效率对比。

**Reply by xbsheng** [reply]
因为执行不同类型计算的时间成本不同，所以通过计算次数衡量时间效率不太科学，我们应该要关心的是“时间”上的快慢

**Reply by Ke-Albert** [reply]
我们常说的大O，就是时间复杂度了吧。相当于总的运行时间=时间复杂度(运行次数)*单位耗时

**Reply by longsizhuo** [reply]
> 我们常说的大O，就是时间复杂度了吧。相当于总的运行时间=时间复杂度(运行次数)*单位耗时

不全对，大O记号 $O(f(n))$ 在数学上表示一个函数的渐近上界。它描述了算法的运行时间（或空间）增长速度最快可以达到 $f(n)$ 的级别，即 $g(n)$ 的增长速度不会超过 $f(n)$。

虽然大O最常用于描述时间复杂度，但它也可以用于描述空间复杂度。

**Comment by hou-jia88** [comment]
总结一下：
       我们在算法（算法与数据结构）的使用过程中，应当主动地甄别能解决同一问题的多种算法，考虑因素有:
一、随着**数据量  or  数据长度  or  数据复杂程度  or  数据类型**的不断增加，该算法在时间资源、空间资源上的利用率
二、对于难以实现实际测试的场景下，进行算法理论估算（复杂度分析），得以最优解之算法。

**Comment by keepil** [comment]
算法可以理解为用最小代价解决问题

**Reply by guoHi21** [reply]
这些代价就包括时间复杂度和空间复杂度还有别的现实代价🥳

**Comment by FVCSAV** [comment]
我觉得这个写的好，好就好在我这种菜狗都看得懂

**Comment by zaozaozuibang** [comment]
效率评估方法主要分为两种：实际测试、理论估算。理论估算是复杂度分析，我们要考虑空间复杂度和时间复杂度。“随着输入数据大小的增加”意味着复杂度反映了算法运行效率与输入数据体量之间的关系。“时间和空间的增长趋势”表示复杂度分析关注的不是运行时间或占用空间的具体值，而是时间或空间增长的“快慢”。

**Comment by W-y-lucy** [comment]
有种像导数，输入变化量△x----->输出的变化量△y，一种相对的概念

**Reply by C7Y-Richer** [reply]
我也是这么理解！

**Comment by RabbitSudo** [comment]
好东西

**Comment by Aka-Edward** [comment]
卧槽！太贴心了还有tips！

**Comment by Quntao10** [comment]
牛逼，目前零基础小白看得津津有味，不觉得枯燥。

**Comment by Atul-8** [comment]
所以算法其实是应用学的理论抽象, 在工程应用学看来只有合适的算法没有垃圾的算法对吧uu🤔

**Comment by xiaqi129** [comment]
是否存在算法销量评估之后，时间复杂度和空间复杂度取舍的情况？
换句话说，是否存在根据实际使用场景，选择牺牲响应时间或者更大内存空间来决策算法的情况？

**Reply by longsizhuo** [reply]
是的,这是工程上的Time-Space Tradeoff. 
一般来说低内存设备（嵌入式）就需要节省空间，内存宝贵，速度次要
而Web API就需要优先速度
如这个网站, 是一个静态网站(评论区除外)那么就空间换时间

**Comment by Gaihuanmingle** [comment]
1

**Comment by SuperZhangRui** [comment]
第二天学习打卡，加油！！！！

**Comment by xiaozhuge123** [comment]
day02

**Comment by zengfantaoGit** [comment]
# day1

**Comment by jidaojiuyou** [comment]
实际上 2C的项目基本上都是空间换时间。甚至上超大规模的分布式系统。

**Reply by jidaojiuyou** [reply]
2C = 面向消费者

**Comment by Xfront-end** [comment]
写跑在开发板上 树莓派这种卡片电脑时， 才能体会到 空间复杂度的重要。

**Comment by EdwinNull** [comment]
打卡

**Comment by Bronc-X** [comment]
## 大家用电脑登着看的还是一边电脑跑代码一边平板看教材

**Comment by jezyzg** [comment]
对复杂度一知半解，不会求

**Reply by John21-create** [reply]
我是先学了数据结构，做个个别算法题，看了作者讲的复杂度的概念后才清楚了很多，可以看看算法题讲解视频，然后再结合书上文字会更好理解

**Comment by pulo2000** [comment]
打卡

**Comment by CrescentZM** [comment]
打卡

**Comment by nimobnaz** [comment]
建议给出简单例子说明复杂度

**Comment by celanwang** [comment]
看到作者对 “实际测试” 和 “理论估算” 的两种方式展开了讨论和分析，这里也有一些自己的感悟。

所谓的实际测试我们有的时候很难去在项目/工程刚刚开始的时候就实现落地，反而理论估算是更加符合在当下软件开发的场景。

举个例子，比如瀑布模型，虽然我们当前不推荐这样的开发模型进行落地，但是它会给我们一些指导建议——我们在开发完成一个工程之前是没有办法看到它的整体完整的一个模型的。

所以这样的话，理论估算是更加重要的，我们可以在工程初期或者是实施阶段，对我们做的东西进行整体上的预估和评价，这样会极大的提高我们对整体上的把握，和降低工程所带来的成本消耗。

所以，我们学习数据结构与算法的本质就是更好的分析和处理，在这个理论阶段进行估算。

当然，算法的本质并不是在实际测试没有任何作用的，而是一个相辅相成的作用，当我们在前期对工程的实现和落地有一个整体的把控之后，对于之后的实际测试也就是一个水到渠成的事情了。

分享链接
https://www.yuque.com/codereview1024/uboc4y/rf2rncqdrgetbu74?singleDoc# 《后端程序员对算法的思考》

**Comment by zhengads** [comment]
好的算法能够在时间和空间复杂度上做到好的平衡，既能够快速出结果，又不会占太大的空间，又快又小是算法设计的根本目的。

**Reply by zhengads** [reply]
当然复杂度分析不只是静态的，更表现了时间和空间随输入规模的增大的变化，有的算法在输入规模比较小时不会让人感觉时间和空间的消耗，但当输入规模一大时，时间和空间的消耗便极为明显。

**Comment by LiLi-0226** [comment]
2.1打卡

**Comment by ttrx2023** [comment]
复杂度分析衡量的就是dy/dx

**Reply by ttrx2023** [reply]
y：时间或空间；     x：输入数据规模

**Comment by Daniel-Zun** [comment]
感觉渐进分析的名字来源于，对大的数据的分析从而取近似值的操作

**Comment by Xiaoha3-3** [comment]
复杂度：数据规模增大，看花费的时间空间的变化趋势。
不过没太懂：”它无需实际运行代码，更加绿色节能。“这个是怎么做到的，看看后面的内容有没有吧~

**Comment by longsizhuo** [comment]
不全对，大O记号 $O(f(n))$ 在数学上表示一个函数的*渐近上界*。它描述了算法的运行时间（或空间）增长速度*最快*可以达到 $f(n)$
的级别，即 $g(n)$ 的增长速度*不会超过* $f(n)$。

虽然大O最常用于描述*时间复杂度*，但它也可以用于描述*空间复杂度*。

Ke-Albert ***@***.***> 于2026年3月20日周五 13:49写道：

**Comment by Ke-Albert** [comment]
没错，这样我想起来了，在cs61b课程中出了Big O，还有一个Big Omega以及其它符号，当时也是听得迷糊




寻梦
***@***.***

**Reply by longsizhuo** [reply]
能帮到你就好

**Comment by hehaibin819** [comment]
厉害厉害

**Comment by xmzd-S** [comment]
渐近复杂度分析（asymptotic complexity analysis）： 描述算法在处理问题时，随着问题规模越来越大时，算法处理问题所需的时间和空间的增长趋势（也就是解决问题所需的时间对数据量的一阶导数）

**Reply by xmzd-S** [reply]
括号里的去掉，不对

**Reply by xmzd-S** [reply]
不用去，是对的
